### DISCUSSION TITLE
Could I use some hooks in Vue Component?

### DISCUSSION THREAD

**Discussion by flyerH** [vikejs/vike#2002] [Discussion] [answered]
For example, I want to report some data before the page jump, since the data depends on some states in the component, if I define the onBeforeRoute.ts file, I have to mount these states on the pageContext so that they can be used in onBeforeRoute.ts. However, this is not very friendly. Is there any solution like the following?
```
import { onBeforeRoute } from 'vike-vue';
```

**Comment by brillout** [comment]
What would this `onBeforeRoute` import do?

**Reply by brillout** [reply]
To answer your question: not currently. (I've a design in mind for programmatically defining hooks, but not sure how relevant it is to the discussion.)

**Reply by flyerH** [reply]
for example
```
onBeforeRoute(() => {
  report({
    componentState1: 'xxxx',
  });
});
```

**Reply by brillout** [reply]
> For example, I want to report some data before the page jump, since the data depends on some states in the component, if I define the onBeforeRoute.ts file, I have to mount these states on the pageContext so that they can be used in onBeforeRoute.ts

Can you elaborate your use case? It isn't clear to me what you're trying to achieve.

**Reply by flyerH** [reply]
```
// component.vue
<template>
  <div>example</div>
</template>

<script lang="ts" setup>
import { onBeforeRoute } from "vike-vue";
import { monitorReport } from "@/util/report";

// the component's own state
const startTime = new Date().getTime();

onBeforeRoute(() => {
  monitorReport({
    page: window.location.href,
    page_from: document.referrer,
    duration: new Date().getTime() - startTime, // report duration before client route change
  });
});
</script>
```
similar #1766

**Reply by brillout** [reply]
Correct me if I'm wrong but AFAICT it's easier if you use `onBeforeRenderClient()`/`onAfterRenderClient()`/`onRenderClient()` instead.

**Reply by flyerH** [reply]
> Correct me if I'm wrong but AFAICT it's easier if you use `onBeforeRenderClient()`/`onAfterRenderClient()`/`onRenderClient()` instead.

In the above example, how do I get startTime in the onBeforeRenderClient method? Is it necessary to set startTime in the pageContext in advance?
If so, is not a very friendly way. I hope to call a hook in the component so that I can use many state in the component directly. Instead of set many component states in the global variable pageContext.

**Reply by brillout** [reply]
I'd use `globalThis`. It seems more natural to me to define this particular user tracking you're describing globally.

If you insist on doing it on a component-level, I think using `usePageContext()` is more natural than your proposed design. You can create a custom hook on top of `usePageContext()` if you want a shared/streamlined DX. I'm curious: isn't that the solution you're looking for? A custom hook? Maybe it's a pattern we should document.

Note that mutating `pageContext` inside components is an officially recommended approach: https://vike.dev/pageContext#custom. (I understand that in theory it isn't the cleanest, but it's the most practical.)

**Reply by flyerH** [reply]
I just found that properties set to pageContext in +Page.vue can't be get in hook file, which seems [by design](https://vike.dev/pageContext#can-i-use-pagecontext-to-store-ui-state). So I can't get component properties from pageContext in the hook during page transitions. I'll try vue-router and see if [onBeforeRouteLeave](https://router.vuejs.org/guide/advanced/composition-api.html#Navigation-Guards) can solve the issue.

**Reply by brillout** [reply]
Have you seen `pageContext._previousPageContext`? Correct me if I'm wrong but AFAICT this gives you the same capabilities than what you'd do with Vue Router.

Keep me updated. I'll update the docs after we resolve the conversation.

**Reply by brillout** [reply]
15cb5132600017fbec629fc6d11423a2af4f8c4b
5e317ec
`vike@0.4.209-commit-5e317ec`

**Reply by flyerH** [reply]
I tried previousPageContext, and it basically solved the issue with hook.
But there is still a small problem: setting variables on usePageContext is only work when the page first visited. If you set variable in page A, and jump to B and then jump back to page A, the custom variable on pageContext will not be set([mini repo](https://github.com/flyerH/vike-vue-example)). Maybe I need to watch page url change and set custom variable when url change.

Anyway, thank you very much for your work, I will give you some sponsorship~

**Reply by brillout** [reply, marked as answer]
> But there is still a small problem: setting variables on usePageContext is only work when the page first visited. If you set variable in page A, and jump to B and then jump back to page A, the custom variable on pageContext will not be set([mini repo](https://github.com/flyerH/vike-vue-example)). Maybe I need to watch page url change and set custom variable when url change.

Seems related to Vue, not to Vike.

> Anyway, thank you very much for your work, I will give you some sponsorship~

Thank you :green_heart:
